### DISCUSSION TITLE
hash_branch routing - routes are not triggered when navigating.

### DISCUSSION THREAD

**Discussion by sumwatt** [jeremyevans/roda#340] [General]
I'm a little stumped and think the answer will be more obvious than I expect. I'm using a slightly modified version of the Roda-Sequel template and modeled my initial routing based on the prefix1 file. 

My default route `/` is pointing to root, no issues loading the page

I added a directory in the routes `./routes/apitest`
I added two files to the routes directory, `apitest.rb` and `products.rb`

I modified `app.rb` to autoload the new directory:

```Ruby
if Unreloader.autoload?
   plugin :autoload_hash_branches
   autoload_hash_branch_dir("./routes")
   autoload_hash_branch_dir("./routes/apitest")
end
```
    

apitest.rb:

```Ruby
class MyTools
  hash_branch("apitest") do |r|
    @nav_items = Family.all
    r.is do
      r.get do
        view "index"
      end
    end
    r.hash_branches
  end
end
```

products.rb:

```Ruby
class MyTools
  hash_branch("/apitest", "products") do |r|
    r.is do
      r.get do
        view "products"
      end
    end
  end
end
```

`/apitest` works fine, but `/apitest/products` does not appear to get loaded/routed (404). I've checked the console and can see `apitest.rb` getting Unloaded/loaded, but `products.rb` does not. I've gone through the documentation and not sure if I'm just overlooking something simple or not. I tried to model my routes similar to those in the Mastering Roda guide but not quite sure if that was the right track. I'd appreciate any pointers!

**Comment by jeremyevans** [comment]
I didn't test, but I'm guessing your problem is with `autoload_hash_branch_dir("./routes/apitest")`, because you are not providing a namespace.  I'm guessing it should be: `autoload_hash_branch_dir("/apitest", "./routes/apitest")`.  With your current code, the autoload_hash_branch plugin would try to load `./routes/apitest/products.rb` if you go to `/products`, since it was setup to look in the default namespace instead of the `/apitest` namespace.   See https://roda.jeremyevans.net/rdoc/classes/Roda/RodaPlugins/AutoloadHashBranches.html

**Comment by sumwatt** [comment]
Thanks for the reply. I think you're pointing me in the right direction. I removed the autoloaders and just did 

```Ruby
Dir["routes/**/*.rb"].each do |route_file|
      require_relative route_file
    end
```
... and changed the directory structure a bit to:
routes: `routes/apitest/products/products.rb`
views: `views/apitest/products/products.erb`

... and the routes are working like I expect. So this points me to something with the autoloader that I'm not doing right. I'll keep poking at it. 

Thanks for your help!

**Comment by opya** [comment]
Hi @sumwatt ,

i tested with [roda-sequel-stack](https://github.com/jeremyevans/roda-sequel-stack) with routes directory structure:

```bash
├── routes
│   ├── apitest
│   │   └── products.rb
│   ├── apitest.rb
│   └── prefix1.rb
```

```ruby
# cat routes/apitest.rb 
class MyApp
  hash_branch("apitest") do |r|
    r.is do
      r.get do
        'APITEST'
      end
    end

    r.hash_branches
  end
end
```

```ruby
# cat routes/apitest/products.rb 
class MyApp
  hash_branch '/apitest', 'products' do |r|
    r.is do
      r.get do
        'PRODUCTS'
      end
    end
  end
end
```

I've added this additional `autoload_hash_branch_dir` for apitest in the app.rb as Jeremy suggested and it worked.

```ruby
autoload_hash_branch_dir('/apitest', './routes/apitest')
```

**Reply by sumwatt** [reply]
@opya - Ah! Thank you. It was the file layout. I originally had it like this:

```
├── routes
│   ├── products.rb
│   ├── apitest.rb
```
... I was misunderstanding where the auto-loader was expecting the child branches, thinking that if the path and namespace were defined it would include the child branches even if they were in the same directory. I added created  `routes/apitest` and moved `products.rb` to it directory and everything works.
